[READY FOR DESIGN REVIEW] -> MachineLearningServices: Adding new API version and definition for AmlInstance compute type

vhamine opened the pull request:
### Latest improvements:
<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow. 
</i><br>
### Contribution checklist:
- [x] I have reviewed the [documentation](https://github.com/Azure/azure-rest-api-specs#basics) for the workflow.
- [x] [Validation tools](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist) were run on swagger spec(s) and have all been fixed in this PR.
- [ ] The [OpenAPI Hub](https://aka.ms/openapiportal) was used for checking validation status and next steps.
### ARM API Review Checklist
- [ ] Service team MUST add the "WaitForARMFeedback" label if the management plane API changes fall into one of the below categories. 
- adding/removing APIs.
- adding/removing properties.
- adding/removing API-version. 
- adding a new service in Azure.

Failure to comply may result in delays for manifest application. Note this does not apply to data plane APIs.
- [ ] If you are blocked on ARM review and want to get the PR merged urgently, please get the ARM oncall for reviews (RP Manifest Approvers team under Azure Resource Manager service) from IcM and reach out to them. 
Please follow the link to find more details on [API review process](https://armwiki.azurewebsites.net/rp_onboarding/ResourceProviderOnboardingAPIRevieworkflow.html).

vhamine committed:
Adding version 2019-11-01

Adding new version 2019-11-01

vhamine committed:
Adding definition for AmlInstance compute type

AutorestCI commented:
# Automation for azure-sdk-for-python

AutorestCI commented:
# Automation for azure-sdk-for-go

azuresdkci commented:
Can one of the admins verify this patch?

vhamine committed:
Modifying examples to use the new API version 2019-11-01

vhamine committed:
Changing readme files to reflect the new API version

vhamine committed:
Bug fixes

vhamine committed:
Bug fix: Adding properties to AmlInstance definitions

vhamine committed:
Bug fix: Adding missing properties in all object types

vhamine committed:
Bug fix: Adding missing properties in startupScript object

Juliehzl commented:
Hi @vhamine, why do you send a PR in draft? Are you ready for review process?

vhamine commented:
> Hi @vhamine, why do you send a PR in draft? Are you ready for review process?

vhamine committed:
Adding POST APIs for start, stop, and restart operations

vhamine commented:
> Hi @vhamine, why do you send a PR in draft? Are you ready for review process?

The PR is not ready for review yet as I am still making changes. This being my first interaction with azure rest API specification, I thought I should create a draft PR and get familiar with the pipeline checks. I plan to attend the office hours today and send a PR for review by EOD. Thanks!

vhamine committed:
Adding state, lastOperation, lastOperationStatus to AmlInstance compute definition

vhamine committed:
Adding examples for start, stop, restart post actions

vhamine committed:
Adding support for Windows OS in AmlCompute compute type

vhamine committed:
Refactoring AmlInstance compute specific properties into separate definitions

vhamine committed:
Adding AmlInstance compute updatable properties to ClusterUpdateProperties definition

vhamine committed:
Adding examples for creating AmlInstance compute

vhamine committed:
Adding example for get AmlInstance

vhamine committed:
Referring to get AmlInstance example in swagger

vhamine committed:
Adding scriptLogLocation as readOnly to custom scripts and updating examples

vhamine committed:
Fixing a bug in latest update

vhamine commented:
This PR achieves the following:
- It defines a new API version 2019-11-01.

- It adds a new ComputeType to MachineLearningServices Workspace RP. The new compute type is called "AmlInstance".

- It defines input specification for the new compute type. This follows the same model as existing compute types as well as same request/response structure.

- It adds Windows OS support for existing AmlCompute compute type.

Juliehzl commented:
> * It adds a new ComputeType to MachineLearningServices Workspace RP. The new compute type is called "AmlInstance".
> *

Thanks for your detailed information. In this way, we can start review process now.

oolukoya commented on specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2019-11-01/machineLearningServices.json:
```diff
@@ -0,0 +1,3060 @@
+      ],
+      "x-ms-discriminator-value": "AKS"
+    },
+    "AmlC
```
We should perhaps add to the description that the default time is 2am. Let me double check on this.

cyly replied:
Could you change it to datastoresMountSettings? In datastore service, they try to use datastore instead of DataStore.

cyly replied:
The same as above. Please apply to all the other changes. Thank you!

cyly replied:
Let's remove the could be a URI. We only need to save the relative path so that later it is easier to do migration.

cyly replied:
If it is maximum is 30 days, why int64? Could we use int32 here?

cyly replied:
The same. If the maximum is 0-23, why int64?

cyly replied:
I think Femi wants to add a version here. Please confirm with Femi.

cyly replied:
I talked with Ahmet and Ahmet suggests not adding this. If needed, we can add it later and it will not be a breaking change. Please remove this property.

vhamine committed:
Adding connectivityEndpoints and createdBy as readOnly properties to AmlInstance spec

vhamine committed:
Removing lastOperation and lastOperationStatus properties from AmlInstance

vhamine committed:
Adding updateVersion property as string to available SDK updates in AmlInstance

vhamine committed:
Removing scriptLogLocation from AmlInstance custom script properties

vhamine committed:
Addressing code review comments

vhamine committed:
Adding adminUserName as readOnly property to AmlInstanceSshSettings

vhamine committed:
Renaming AmlInstance to ComputeInstance

vhamine committed:
Adding instanceSharing property to ComputeInstance

cyly commented on specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2019-11-01/machineLearningServices.json:
```diff
@@ -1978,6 +1978,146 @@
+                  },
+                  "adminCredentials": {
+                    "type": "object",
+                    "adminPublicKey": {
```
Do you need to follow the convention to define publickey in the properties. Also as what Chris said, you need to add a readonly property for the admin name. Otherwise there is no way to tell users the admin user name.

vhamine committed:
Internal review feedback: Adding sshPort to sshSettings and separating applicationURIs from connectivityEndpoints

vhamine committed:
Fixing SpellCheck error

vhamine committed:
Keeping minimal set of properties (for Ignite) in ComputeInstance spec

cyly commented on specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2019-11-01/examples/createComputeInstance.json:
```diff
@@ -0,0 +1,52 @@
+        "properties": {
+          "vmSize": "STANDARD_NC6",
+          "subnet": "test-subnet-resource-id",
+          "applicationSharingPolicy": "Personal",
```
I didn't see this defined in the amlInstance request and response. Do you miss that in the definition?

vhamine replied:
Yu, as discussed, you were not looking at the latest version.

vhamine committed:
Renaming virtualMachineImageId to virtualMachineImage

sanjaiganesh commented on specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2019-11-01/machineLearningServices.json:
```diff
@@ -0,0 +1,2927 @@
+      ],
+      "x-ms-discriminator-value": "AKS"
+    },
+    "AmlCom
```
>"osType": { [](start = 16, length = 11)

add default 

 "default": Linux

sanjaiganesh replied:
>applicationUris [](start = 17, length = 15)

rename.. applicationUris => applications

sanjaiganesh replied:
>ComputeInstanceApplicationUri [](start = 5, length = 29)

rename : ComputeInstanceApplicationUri => ComputeInstanceApplication

sanjaiganesh commented on specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2019-11-01/machineLearningServices.json:
```diff
@@ -0,0 +1,2927 @@
+      ],
+      "x-ms-discriminator-value": "AKS"
+    },
+    "AmlCom
```
strip off application from applicationName and applicationUri  

and consider adding icon for display purposes.. having this in mind, i would suggest

"displayName"
"endpointUri"
"displayIcon" [future]

vhamine replied:
Thanks, done!

vhamine committed:
Addressing ARM review feedback

vhamine committed:
Making Linux the default OS type in AmlCompute

vhamine commented:
@Juliehzl this PR was signed-off by ARM. We do not want to merge this PR yet since our back-end is still being implemented. We should co-ordinate merging this before Ignite 2019. Can you please let us know what the next steps should be?

CC: @ahcan76

Juliehzl commented on specification/machinelearningservices/resource-manager/Microsoft.MachineLearningServices/stable/2019-11-01/machineLearningServices.json:
```diff
@@ -0,0 +1,2928 @@
+          }
+        },
+        "x-ms-pageable": {
+          "nextLinkName": "nextLink"
```
Where is your nextLink property for nextLinkName?

Juliehzl commented:
Hi @vhamine, I have a look at your swagger specification. It looks good to me and I already approved for your PR. All review process is completed before clicking the merge button.
If you don't want it to be merged now, I will add DONOTMERGE label for your PR and when you are ready to merge, just let me know and I will click the merge button for you.

ramagott committed:
Updating the ComputeInstance state list

1. Updating the ComputeInstance state list
2. Changing the return type for CI start, stop, and restart operations

ramagott committed:
Updating return type in examples for CI start, stop and restart operations

ramagott committed:
Fixing braces

akning-ms commented:
Hi @vhamine, Can we know what is this PR status as it is existing for a long time？

ramagott commented:
> Hi @vhamine, Can we know what is this PR status as it is existing for a long time？

This PR is not required any more. Can you close it?
